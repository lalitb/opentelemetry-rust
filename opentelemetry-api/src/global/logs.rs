use std::{
    borrow::Cow,
    fmt, mem,
    sync::{Arc, RwLock},
    time::SystemTime,
};

use once_cell::sync::Lazy;

use crate::{
    Key,
    logs::{AnyValue, Logger, LogRecord, LoggerProvider, NoopLoggerProvider, Severity},
    trace::SpanContext,
    InstrumentationLibrary
};

/// Allows a specific [`crate::trace::LogRecord`] to be used generically by [`BoxedLogRecord`]
/// instances by mirroring the interface and boxing the return types.
pub trait ObjectSafeLogRecord {
    fn with_timestamp(self, timestamp: SystemTime);

    fn with_observed_timestamp(self, timestamp: SystemTime);

    fn with_span_context(self, span_context: &SpanContext);

    fn with_severity_text(self, severity_text: Cow<'static, str>);

    fn with_severity_number(self, severity_number: Severity);

    fn with_body(self, body: AnyValue);

    fn with_attributes(self, attributes: Vec<(Key, AnyValue)>);

    fn with_attribute(self, key: Key, value: AnyValue);
}

impl<L: LogRecord> ObjectSafeLogRecord for L
{
    fn with_timestamp(self, timestamp: SystemTime) {
        self.with_timestamp(timestamp)
    }

    fn with_observed_timestamp(self, timestamp: SystemTime) {
        self.with_observed_timestamp(timestamp)
    }

    fn with_span_context(self, span_context: &SpanContext) {
        self.with_span_context(span_context)
    }

    fn with_severity_text(self, severity_text: Cow<'static, str>)
     {
        self.with_severity_text(severity_text)
    }

    fn with_severity_number(self, severity_number: Severity) {
        self.with_severity_number(severity_number)
    }

    fn with_body(self, body: AnyValue) {
        self.with_body(body)
    }

    fn with_attributes(self, attributes: Vec<(Key, AnyValue)>) {
        self.with_attributes(attributes)
    }

    fn with_attribute(self, key: Key, value: AnyValue)
    {
        self.with_attribute(key, value)
    }
}

pub struct BoxedLogRecord(Box<dyn ObjectSafeLogRecord + Send + Sync + 'static>);

impl BoxedLogRecord {
    pub(crate) fn new<L>(record: L) -> Self 
    where
        L: ObjectSafeLogRecord + Send + Sync + 'static,
    {
        BoxedLogRecord(Box::new(record))
    }

    pub fn into_inner(self) -> Box<dyn ObjectSafeLogRecord + Send + Sync + 'static> {
        self.0
    }
}


impl fmt::Debug for BoxedLogRecord {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("BoxedLogRecord")
    }
}

impl LogRecord for BoxedLogRecord {

    fn with_timestamp(self, timestamp: SystemTime) {
        self.0.with_timestamp(timestamp)
    }

    fn with_observed_timestamp(self, timestamp: SystemTime) {
        self.0.with_observed_timestamp(timestamp)
    }

    fn with_span_context(self, span_context: &SpanContext) {
        self.0.with_span_context(span_context)
    }

    fn with_severity_text(self, severity_text: Cow<'static, str>) {
        self.0.with_severity_text(severity_text)
    }

    fn with_severity_number(self, severity_number: Severity) {
        self.0.with_severity_number(severity_number)
    }

    fn with_body(self, body: AnyValue) {
        self.0.with_body(body)
    }

    fn with_attributes(self, attributes: Vec<(Key, AnyValue)>) {
        self.0.with_attributes(attributes)
    }

    fn with_attribute(self, key: Key, value: AnyValue)

    {
        self.0.with_attribute(key, value)
    }

}

/// Wraps the [`GlobalLoggerProvider`]'s [`Logger`] so it can be used generically by
/// applications without knowing the underlying type.

pub struct BoxedLogger(Box<dyn ObjectSafeLogger + Send + Sync>);

impl BoxedLogger {
    /// Create a `BoxedTracer` from an object-safe tracer.
    pub fn new(logger: Box<dyn ObjectSafeLogger + Send + Sync>) -> Self {
        BoxedLogger(logger)
    }
}

impl fmt::Debug for BoxedLogger {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("BoxedLogger")
    }
}

impl Logger for BoxedLogger {
    /// Global logger uses `BoxedLogRecord`s so that it can be a global singleton,
    /// which is not possible if it takes generic type parameters.
    type LogRecordType = BoxedLogRecord;

    fn create_log_record(&self) -> Self::LogRecordType {
        BoxedLogRecord(self.0.create_log_record_boxed())
    }

    fn emit(&self, record: Self::LogRecordType) {
        self.0.emit(record.into_inner())
    }
}

/// Allows a specific [`Logger`] to be used generically by [`BoxedLogger`]
/// instances by mirroring the interface and boxing the return types.
///
/// [`Logger`]: crate::logs::Logger
/// 

pub trait ObjectSafeLogger {

    /// Creates a new [`LogRecord`] that is a trait object through the underlying
    /// [`Logger`].
    ///
    /// [`LogRecord`]: crate::logs::LogRecord
    /// [`Logger`]: crate::logs::Logger
    fn create_log_record_boxed(&self) -> Box<dyn ObjectSafeLogRecord + Send + Sync>;

    /// Emits a [`LogRecord`] that is a trait object through the underlying
    /// 
    fn emit(&self, record: Box<dyn ObjectSafeLogRecord + Send + Sync>);

}

impl<R, L> ObjectSafeLogger for L
where
    R: LogRecord + Send + Sync + 'static,   
    L: Logger<LogRecordType = R>
{  

    fn create_log_record_boxed(&self) ->  Box<dyn ObjectSafeLogRecord + Send + Sync> {
        Box::new(self.create_log_record())
    }

    fn emit(&self, record: Box<dyn ObjectSafeLogRecord + Send + Sync>) {
        self.emit(*record)
    }

}


/// Allows a specific [`LoggerProvider`] to be used generically, by mirroring
/// the interface, and boxing the returned types.
///
/// [`LoggerProvider`]: crate::logs::LoggerProvider.
pub trait ObjectSafeLoggerProvider {
    /// Creates a versioned named [`Logger`] instance that is a trait object
    /// through the underlying [`LoggerProvider`].
    ///
    /// [`Logger`]: crate::logs::Logger
    /// [`LoggerProvider`]: crate::logs::LoggerProvider
    fn boxed_logger(
        &self,
        library: Arc<InstrumentationLibrary>,
    ) -> Box<dyn ObjectSafeLogger + Send + Sync + 'static>;
}

impl<R, L, P> ObjectSafeLoggerProvider for P
where
    R: LogRecord + Send + Sync + 'static,
    L: Logger<LogRecordType = R> + Send + Sync + 'static,
    P: LoggerProvider<Logger = L>,
{
    fn boxed_logger(
        &self,
        library: Arc<InstrumentationLibrary>,
    ) -> Box<dyn ObjectSafeLogger + Send + Sync + 'static> {
        Box::new(self.library_logger(library))
    }
}

#[derive(Clone)]
/// Represents the globally configured [`LoggerProvider`] instance.
pub struct GlobalLoggerProvider {
    provider: Arc<dyn ObjectSafeLoggerProvider + Send + Sync>,
}

impl fmt::Debug for GlobalLoggerProvider {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("GlobalLoggerProvider")
    }
}

impl GlobalLoggerProvider {
    fn new<
        R: LogRecord + Send + Sync + 'static,
        L: Logger + Send + Sync + 'static,
        P: LoggerProvider<Logger = L> + Send + Sync + 'static,
    >(
        provider: P,
    ) -> Self {
        GlobalLoggerProvider {
            provider: Arc::new(provider),
        }
    }
}

impl LoggerProvider for GlobalLoggerProvider {
    type Logger = BoxedLogger;

    fn library_logger(&self, library: Arc<InstrumentationLibrary>) -> Self::Logger {
        BoxedLogger(self.provider.boxed_logger(library))
    }
}

static GLOBAL_LOGGER_PROVIDER: Lazy<RwLock<GlobalLoggerProvider>> =
    Lazy::new(|| RwLock::new(GlobalLoggerProvider::new(NoopLoggerProvider::new())));

/// Returns an instance of the currently configured global [`LoggerProvider`]
/// through [`GlobalLoggerProvider`].
///
/// [`LoggerProvider`]: crate::logs::LoggerProvider
pub fn logger_provider() -> GlobalLoggerProvider {
    GLOBAL_LOGGER_PROVIDER
        .read()
        .expect("GLOBAL_LOGGER_PROVIDER RwLock poisoned")
        .clone()
}

/// Creates a named instance of [`Logger`] via the configured
/// [`GlobalLoggerProvider`].
///
/// If `name` is an empty string, the provider will use a default name.
///
/// [`Logger`]: crate::logs::Logger
pub fn logger(name: Cow<'static, str>) -> BoxedLogger {
    logger_provider().logger(name)
}

/// Sets the given [`LoggerProvider`] instance as the current global provider,
/// returning the [`LoggerProvider`] instance that was previously set as global
/// provider.
pub fn set_logger_provider<L, P>(new_provider: P) -> GlobalLoggerProvider
where
    L: Logger + Send + Sync + 'static,
    P: LoggerProvider<Logger = L> + Send + Sync + 'static,
{
    let mut provider = GLOBAL_LOGGER_PROVIDER
        .write()
        .expect("GLOBAL_LOGGER_PROVIDER RwLock poisoned");
    mem::replace(&mut *provider, GlobalLoggerProvider::new(new_provider))
}

/// Shut down the current global [`LoggerProvider`].
pub fn shutdown_logger_provider() {
    let _ = set_logger_provider(NoopLoggerProvider::new());
}
