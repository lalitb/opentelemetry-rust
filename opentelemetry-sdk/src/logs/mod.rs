//! # OpenTelemetry Log SDK

mod config;
mod log_emitter;
mod log_processor;

pub use config::Config;
pub use log_emitter::{Builder, Logger, LoggerProvider};
pub use log_processor::{
    BatchConfig, BatchLogProcessor, BatchLogProcessorBuilder, BatchMessage, LogProcessor,
    SimpleLogProcessor,
};
pub use opentelemetry_api::logs::Severity;
use std::fmt;

/// Interface for checking if a log level is enabled.
pub trait EventEnabled: Send + Sync + fmt::Debug {
    /// Generate a new `TraceId`
    fn is_enabled(&self, _name: &str, _level: Severity) -> bool;
}

/// Default [`EventEnabled`] implementation.
///
/// return true for all log levels.
#[derive(Clone, Debug, Default)]
pub struct DefaultEventEnabled {
    _private: (),
}

impl EventEnabled for DefaultEventEnabled {
    fn is_enabled(&self, _name: &str, _level: Severity) -> bool {
        true
    }
}
