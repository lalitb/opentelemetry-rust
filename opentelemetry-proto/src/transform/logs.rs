#[cfg(feature = "gen-tonic-messages")]
pub mod tonic {
    use crate::{
        tonic::{
            common::v1::{any_value::Value, AnyValue, ArrayValue, KeyValue, KeyValueList, InstrumentationScope},
            logs::v1::{LogRecord, ResourceLogs, ScopeLogs, SeverityNumber},
            resource::v1::Resource,
            Attributes,
        },
        transform::common::{to_nanos, tonic::ResourceAttributesWithSchema},
    };
    use opentelemetry::logs::{AnyValue as LogsAnyValue, Severity};
    use std::collections::HashMap;

    impl From<LogsAnyValue> for AnyValue {
        fn from(value: LogsAnyValue) -> Self {
            AnyValue {
                value: Some(value.into()),
            }
        }
    }

    impl From<LogsAnyValue> for Value {
        fn from(value: LogsAnyValue) -> Self {
            match value {
                LogsAnyValue::Double(f) => Value::DoubleValue(f),
                LogsAnyValue::Int(i) => Value::IntValue(i),
                LogsAnyValue::String(s) => Value::StringValue(s.into()),
                LogsAnyValue::Boolean(b) => Value::BoolValue(b),
                LogsAnyValue::ListAny(v) => Value::ArrayValue(ArrayValue {
                    values: v
                        .into_iter()
                        .map(|v| AnyValue {
                            value: Some(v.into()),
                        })
                        .collect(),
                }),
                LogsAnyValue::Map(m) => Value::KvlistValue(KeyValueList {
                    values: m
                        .into_iter()
                        .map(|(key, value)| KeyValue {
                            key: key.into(),
                            value: Some(AnyValue {
                                value: Some(value.into()),
                            }),
                        })
                        .collect(),
                }),
                LogsAnyValue::Bytes(v) => Value::BytesValue(v),
            }
        }
    }

    impl From<opentelemetry_sdk::logs::LogRecord> for LogRecord {
        fn from(log_record: opentelemetry_sdk::logs::LogRecord) -> Self {
            let trace_context = log_record.trace_context.as_ref();
            let severity_number = match log_record.severity_number {
                Some(Severity::Trace) => SeverityNumber::Trace,
                Some(Severity::Trace2) => SeverityNumber::Trace2,
                Some(Severity::Trace3) => SeverityNumber::Trace3,
                Some(Severity::Trace4) => SeverityNumber::Trace4,
                Some(Severity::Debug) => SeverityNumber::Debug,
                Some(Severity::Debug2) => SeverityNumber::Debug2,
                Some(Severity::Debug3) => SeverityNumber::Debug3,
                Some(Severity::Debug4) => SeverityNumber::Debug4,
                Some(Severity::Info) => SeverityNumber::Info,
                Some(Severity::Info2) => SeverityNumber::Info2,
                Some(Severity::Info3) => SeverityNumber::Info3,
                Some(Severity::Info4) => SeverityNumber::Info4,
                Some(Severity::Warn) => SeverityNumber::Warn,
                Some(Severity::Warn2) => SeverityNumber::Warn2,
                Some(Severity::Warn3) => SeverityNumber::Warn3,
                Some(Severity::Warn4) => SeverityNumber::Warn4,
                Some(Severity::Error) => SeverityNumber::Error,
                Some(Severity::Error2) => SeverityNumber::Error2,
                Some(Severity::Error3) => SeverityNumber::Error3,
                Some(Severity::Error4) => SeverityNumber::Error4,
                Some(Severity::Fatal) => SeverityNumber::Fatal,
                Some(Severity::Fatal2) => SeverityNumber::Fatal2,
                Some(Severity::Fatal3) => SeverityNumber::Fatal3,
                Some(Severity::Fatal4) => SeverityNumber::Fatal4,
                None => SeverityNumber::Unspecified,
            };

            LogRecord {
                time_unix_nano: log_record.timestamp.map(to_nanos).unwrap_or_default(),
                observed_time_unix_nano: to_nanos(log_record.observed_timestamp.unwrap()),
                severity_number: severity_number.into(),
                severity_text: log_record.severity_text.map(Into::into).unwrap_or_default(),
                body: log_record.body.map(Into::into),
                attributes: log_record
                    .attributes
                    .map(Attributes::from_iter)
                    .unwrap_or_default()
                    .0,
                dropped_attributes_count: 0,
                flags: trace_context
                    .map(|ctx| {
                        ctx.trace_flags
                            .map(|flags| flags.to_u8() as u32)
                            .unwrap_or_default()
                    })
                    .unwrap_or_default(),
                span_id: trace_context
                    .map(|ctx| ctx.span_id.to_bytes().to_vec())
                    .unwrap_or_default(),
                trace_id: trace_context
                    .map(|ctx| ctx.trace_id.to_bytes().to_vec())
                    .unwrap_or_default(),
            }
        }
    }

    impl
        From<(
            opentelemetry_sdk::export::logs::LogData,
            &ResourceAttributesWithSchema,
        )> for ResourceLogs
    {
        fn from(
            data: (
                opentelemetry_sdk::export::logs::LogData,
                &ResourceAttributesWithSchema,
            ),
        ) -> Self {
            let (log_data, resource) = data;

            ResourceLogs {
                resource: Some(Resource {
                    attributes: resource.attributes.0.clone(),
                    dropped_attributes_count: 0,
                }),
                schema_url: resource.schema_url.clone().unwrap_or_default(),
                scope_logs: vec![ScopeLogs {
                    schema_url: log_data
                        .instrumentation
                        .schema_url
                        .clone()
                        .map(Into::into)
                        .unwrap_or_default(),
                    scope: Some(log_data.instrumentation.into()),
                    log_records: vec![log_data.record.into()],
                }],
            }
        }
    }

    pub fn group_logs_by_resource_and_scope<'a>(logs: Vec<opentelemetry_sdk::export::logs::LogData>, resource: &ResourceAttributesWithSchema) -> Vec<ResourceLogs> {
        // no need of explicit grouping by resource, as all the logs belong to single resource.
        let scope_map = logs.iter().fold(
            HashMap::new(),
            |mut scope_map: HashMap<InstrumentationScopeKey, (&'a [opentelemetry_proto::tonic::common::v1::KeyValue], Vec<&LogData>)>, log| {
                let instrumentation = &log.instrumentation;
                let scope_key = InstrumentationScopeKey::from(instrumentation);
                let entry = scope_map.entry(scope_key).or_insert_with(|| (instrumentation.attributes.as_slice(), Vec::new()));
                entry.1.push(log);
                scope_map
            },
        );
    
        let scope_logs = scope_map.into_iter().map(|(scope_key, (attributes, log_records))| {
            ScopeLogs {
                scope: Some(InstrumentationScope {
                    name: scope_key.name.to_string(),
                    version: scope_key.version.unwrap_or_default().to_string(),
                    schema_url: scope_key.schema_url.unwrap_or_default().to_string(),
                    attributes: attributes.to_vec(),
                    dropped_attributes_count: 0, // Assuming no dropped attributes
                }),
                log_records: log_records.into_iter().map(|log| log.clone().into()).collect(),
                ..Default::default()
            }
        }).collect();
    
        ResourceLogs {
            resource: Some(opentelemetry_proto::tonic::resource::v1::Resource {
                attributes: resource.attributes.0.clone(),
                dropped_attributes_count: 0,
            }),
            scope_logs,
            schema_url: resource.schema_url.clone().unwrap_or_default(),
        }
    }
}
