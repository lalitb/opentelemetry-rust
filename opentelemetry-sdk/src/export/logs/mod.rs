//! Log exporters
use crate::logs::LogRecord;
use crate::Resource;
use async_trait::async_trait;
#[cfg(feature = "logs_level_enabled")]
use opentelemetry::logs::Severity;
use opentelemetry::{
    logs::{LogError, LogResult},
    InstrumentationLibrary,
};
use std::fmt::Debug;

/// A batch of log records to be exported by a `LogExporter`.
#[derive(Debug)]
pub struct LogBatch<'a> {
    data: &'a [(&'a LogRecord, &'a InstrumentationLibrary)],
}

impl<'a> LogBatch<'a> {
    pub fn new(data: &'a [(&'a LogRecord, &'a InstrumentationLibrary)]) -> LogBatch<'a> {
        LogBatch { data }
    }
}

impl LogBatch<'_> {
    /// iterator
    pub fn iter(&self) -> impl Iterator<Item = (&LogRecord, &InstrumentationLibrary)> {
        self.data
            .iter()
            .map(|(record, library)| (*record, *library))
    }
}

/// `LogExporter` defines the interface that log exporters should implement.
#[async_trait]
pub trait LogExporter: Send + Sync + Debug {
    /// Exports a batch of [`LogRecord`, `InstrumentationLibrary`].
    async fn export(&mut self, batch: LogBatch<'_>) -> LogResult<()>;
    /// Shuts down the exporter.
    fn shutdown(&mut self) {}
    #[cfg(feature = "logs_level_enabled")]
    /// Chek if logs are enabled.
    fn event_enabled(&self, _level: Severity, _target: &str, _name: &str) -> bool {
        // By default, all logs are enabled
        true
    }
    /// Set the resource for the exporter.
    fn set_resource(&mut self, _resource: &Resource) {}
}

/// Describes the result of an export.
pub type ExportResult = Result<(), LogError>;
