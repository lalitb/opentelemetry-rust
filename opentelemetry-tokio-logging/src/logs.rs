use std::marker;

use opentelemetry_sdk::logs::{Logger, LoggerProvider};
use opentelemetry_api::{
    logs as otel,
    logs::{
        LogRecord, LogResult, 
    },
    Context as OtelContext,
};
