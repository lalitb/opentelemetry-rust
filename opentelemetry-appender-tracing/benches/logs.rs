use async_trait::async_trait;
use criterion::{criterion_group, criterion_main, Criterion};
use opentelemetry::logs::LogResult;
use opentelemetry::KeyValue;
use opentelemetry_appender_tracing::layer as tracing_layer;
use opentelemetry_sdk::export::logs::{LogData, LogExporter};
use opentelemetry_sdk::logs::{Config, LogProcessor, LoggerProvider};
use opentelemetry_sdk::Resource;
use tracing::info;
use tracing_subscriber::prelude::*;
use tracing_subscriber::Registry;

#[derive(Debug, Clone)]
struct NoopExporter {
    enabled: bool,
}

#[async_trait]
impl LogExporter for NoopExporter {
    async fn export(&mut self, _: Vec<LogData>) -> LogResult<()> {
        LogResult::Ok(())
    }

    fn event_enabled(&self, _: opentelemetry::logs::Severity, _: &str, _: &str) -> bool {
        self.enabled
    }
}

#[derive(Debug)]
struct NoopProcessor {
    exporter: Box<dyn LogExporter>,
}

impl NoopProcessor {
    fn new(exporter: Box<dyn LogExporter>) -> Self {
        Self { exporter }
    }
}

impl LogProcessor for NoopProcessor {
    fn emit(&self, _: LogData) {
        // no-op
    }

    fn force_flush(&self) -> LogResult<()> {
        Ok(())
    }

    fn shutdown(&mut self) -> LogResult<()> {
        Ok(())
    }

    fn event_enabled(
        &self,
        level: opentelemetry::logs::Severity,
        target: &str,
        name: &str,
    ) -> bool {
        self.exporter.event_enabled(level, target, name)
    }
}

fn benchmark_no_subscriber(c: &mut Criterion) {
    c.bench_function("log_no_subscriber", |b| {
        b.iter(|| {
            info!("my-event-name");
        });
    });
}

fn benchmark_with_subscriber(c: &mut Criterion, enabled: bool, bench_name: &str) {
    let exporter = NoopExporter { enabled };
    let processor = NoopProcessor::new(Box::new(exporter));
    let provider = LoggerProvider::builder()
        .with_config(
            Config::default().with_resource(Resource::new(vec![KeyValue::new(
                "service.name",
                "log-appender-tracing-example",
            )])),
        )
        .with_log_processor(processor)
        .build();
    let tracing_layer = tracing_layer::OpenTelemetryTracingBridge::new(&provider);
    let subscriber = Registry::default().with(tracing_layer);

    tracing::subscriber::with_default(subscriber, || {
        c.bench_function(bench_name, |b| {
            b.iter(|| {
                info!("my-event-name");
            });
        });
    });
}

fn criterion_benchmark(c: &mut Criterion) {
    benchmark_no_subscriber(c);
    benchmark_with_subscriber(c, false, "log_subscriber_disabled");
    benchmark_with_subscriber(c, true, "log_subscriber_enabled");
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
