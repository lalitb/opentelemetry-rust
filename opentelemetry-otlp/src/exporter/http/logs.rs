use std::sync::Arc;

use async_trait::async_trait;
use http::{header::CONTENT_TYPE, Method};
use opentelemetry::logs::{LogError, LogResult};
use opentelemetry_sdk::export::logs::{LogData, LogExporter};

use super::OtlpHttpClient;

#[async_trait]
impl LogExporter for OtlpHttpClient {
    async fn export(&mut self, batch: Vec<LogData>) -> LogResult<()> {
        let client = self
            .client
            .lock()
            .map_err(|e| LogError::Other(e.to_string().into()))
            .and_then(|g| match &*g {
                Some(client) => Ok(Arc::clone(client)),
                _ => Err(LogError::Other("exporter is already shut down".into())),
            })?;

        let (body, content_type) = {
            let resource = self.resource.lock().unwrap();
            build_body(batch, &resource)?
        };
        let mut request = http::Request::builder()
            .method(Method::POST)
            .uri(&self.collector_endpoint)
            .header(CONTENT_TYPE, content_type)
            .body(body)
            .map_err(|e| crate::Error::RequestFailed(Box::new(e)))?;

        for (k, v) in &self.headers {
            request.headers_mut().insert(k.clone(), v.clone());
        }

        let request_uri = request.uri().to_string();
        let response = client.send(request).await?;

        if !response.status().is_success() {
            let error = format!(
                "OpenTelemetry logs export failed. Url: {}, Status Code: {}, Response: {:?}",
                response.status().as_u16(),
                request_uri,
                response.body()
            );
            return Err(LogError::Other(error.into()));
        }

        Ok(())
    }

    fn shutdown(&mut self) {
        let _ = self.client.lock().map(|mut c| c.take());
    }

    fn set_resource(&mut self, resource: &opentelemetry_sdk::Resource) {
        let mut res = self.resource.lock().unwrap();
        *res = resource.clone();
    }
}

#[cfg(feature = "http-proto")]
fn build_body(
    logs: Vec<LogData>,
    resource: &opentelemetry_sdk::Resource,
) -> LogResult<(Vec<u8>, &'static str)> {
    use opentelemetry_proto::tonic::collector::logs::v1::ExportLogsServiceRequest;
    use prost::Message;
    let resource_logs = logs
        .into_iter()
        .map(|log_event| (log_event, resource).into())
        .collect::<Vec<_>>();

    let req = ExportLogsServiceRequest { resource_logs };
    let mut buf = vec![];
    req.encode(&mut buf).map_err(crate::Error::from)?;

    Ok((buf, "application/x-protobuf"))
}

#[cfg(not(feature = "http-proto"))]
fn build_body(
    logs: Vec<LogData>,
    resource: &opentelemetry_sdk::Resource,
) -> LogResult<(Vec<u8>, &'static str)> {
    Err(LogsError::Other(
        "No http protocol configured. Enable one via `http-proto`".into(),
    ))
}
