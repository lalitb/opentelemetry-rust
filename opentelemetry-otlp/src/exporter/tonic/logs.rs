use async_trait::async_trait;
use core::fmt;
use opentelemetry_proto::tonic::collector::logs::v1::{
    logs_service_client::LogsServiceClient, ExportLogsServiceRequest,
};
use opentelemetry_sdk::export::logs::{LogBatch, LogExporter};
use opentelemetry_sdk::logs::{LogError, LogResult};
use tonic::{codegen::CompressionEncoding, service::Interceptor, transport::Channel, Request};

use opentelemetry_proto::transform::logs::tonic::group_logs_by_resource_and_scope;

use super::BoxInterceptor;
use std::sync::atomic::{AtomicBool, Ordering};

pub(crate) struct TonicLogsClient {
    inner: ClientInner,
    is_shutdown: AtomicBool,
    #[allow(dead_code)]
    // <allow dead> would be removed once we support set_resource for metrics.
    resource: opentelemetry_proto::transform::common::tonic::ResourceAttributesWithSchema,
}

struct ClientInner {
    client: Mutex<LogsServiceClient<Channel>>,
    interceptor: BoxInterceptor,
}

impl fmt::Debug for TonicLogsClient {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("TonicLogsClient")
    }
}

impl TonicLogsClient {
    pub(super) fn new(
        channel: Channel,
        interceptor: BoxInterceptor,
        compression: Option<CompressionEncoding>,
    ) -> Self {
        let mut client = LogsServiceClient::new(channel);
        if let Some(compression) = compression {
            client = client
                .send_compressed(compression)
                .accept_compressed(compression);
        }

        TonicLogsClient {
            inner: ClientInner {
                client,
                interceptor,
            },
            is_shutdown: AtomicBool::new(false),
            resource: Default::default(),
        }
    }
}

#[async_trait]
impl LogExporter for TonicLogsClient {
    async fn export(&self, batch: LogBatch<'_>) -> LogResult<()> {
        if self.is_shutdown.load(Ordering::SeqCst) {
            return Err(LogError::Other("exporter is already shut down".into()));
        }

        let (metadata, extensions, _) = self
            .inner
            .interceptor
            .call(Request::new(()))
            .map_err(|e| LogError::Other(Box::new(e)))?
            .into_parts();

        let resource_logs = group_logs_by_resource_and_scope(batch, &self.resource);

        self.inner
            .client
            .export(Request::from_parts(
                metadata,
                extensions,
                ExportLogsServiceRequest { resource_logs },
            ))
            .await
            .map_err(crate::Error::from)?;

        Ok(())
    }

    fn shutdown(&mut self) {}

    fn set_resource(&mut self, resource: &opentelemetry_sdk::Resource) {
        self.resource = resource.into();
    }
}
