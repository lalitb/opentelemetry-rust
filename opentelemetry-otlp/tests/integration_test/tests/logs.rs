#![cfg(unix)]

use integration_test_runner::logs_asserter::{read_logs_from_json, LogsAsserter};
use integration_test_runner::Protocol;
use log::{info, Level, Log, SetLoggerError};
use opentelemetry::logs::LogError;
use opentelemetry::KeyValue;
use opentelemetry_appender_log::OpenTelemetryLogBridge;
use opentelemetry_sdk::{logs as sdklogs, runtime, Resource};
use std::error::Error;
use std::fs::File;
use std::os::unix::fs::MetadataExt;

use lazy_static::lazy_static;
use std::sync::Mutex;

lazy_static! {
    static ref LOGGER: Mutex<Option<Box<dyn Log>>> = Mutex::new(None);
}

fn init_logger(logger: Box<dyn Log>) -> Result<(), SetLoggerError> {
    let mut global_logger = LOGGER.lock().unwrap();
    if global_logger.is_none() {
        *global_logger = Some(logger);
        log::set_logger(global_logger.as_ref().unwrap().as_ref())
    } else {
        Ok(())
    }
}

trait ExporterBuilder: Send + Sync {
    fn build(self: Box<Self>) -> Result<sdklogs::LoggerProvider, LogError>;
}

impl ExporterBuilder for opentelemetry_otlp::TonicExporterBuilder {
    fn build(self: Box<Self>) -> Result<sdklogs::LoggerProvider, LogError> {
        opentelemetry_otlp::new_pipeline()
            .logging()
            .with_exporter(*self)
            .with_resource(Resource::new(vec![KeyValue::new(
                opentelemetry_semantic_conventions::resource::SERVICE_NAME,
                "logs-integration-test",
            )]))
            .install_batch(runtime::Tokio)
    }
}

impl ExporterBuilder for opentelemetry_otlp::HttpExporterBuilder {
    fn build(self: Box<Self>) -> Result<sdklogs::LoggerProvider, LogError> {
        opentelemetry_otlp::new_pipeline()
            .logging()
            .with_exporter(*self)
            .with_resource(Resource::new(vec![KeyValue::new(
                opentelemetry_semantic_conventions::resource::SERVICE_NAME,
                "logs-integration-test",
            )]))
            .install_batch(runtime::Tokio)
    }
}

fn init_logs(protocol: &Protocol) -> Result<Box<dyn ExporterBuilder>, LogError> {
    let exporter: Box<dyn ExporterBuilder> = match protocol {
        Protocol::Tonic => Box::new(opentelemetry_otlp::new_exporter().tonic()),
        Protocol::Http => Box::new(opentelemetry_otlp::new_exporter().http()),
    };

    Ok(exporter)
}

pub async fn logs(protocol: &Protocol) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
    let exporter = init_logs(protocol)?;
    println!("init done..");
    let logger_provider = exporter.build()?;
    println!("build done..");
    let otel_log_appender = OpenTelemetryLogBridge::new(&logger_provider);
    println!("bridge done..");
    log::set_boxed_logger(Box::new(otel_log_appender))?;
    println!("set logger done..");
    log::set_max_level(Level::Info.to_level_filter());
    println!("let's log..");
    info!(target: "my-target", "hello from {}. My price is {}.", "banana", 2.99);
    println!("log done..");
    let _ = logger_provider.shutdown();
    println!("shutdown done..");
    Ok(())
}

pub fn assert_logs_results(result: &str, expected: &str) {
    let left = read_logs_from_json(File::open(expected).unwrap());
    let right = read_logs_from_json(File::open(result).unwrap());

    LogsAsserter::new(left, right).assert();

    assert!(File::open(result).unwrap().metadata().unwrap().size() > 0)
}

#[test]
#[should_panic(expected = "assertion `left == right` failed: body does not match")]
pub fn test_assert_logs_eq_failure() {
    let left = read_logs_from_json(File::open("./expected/logs.json").unwrap());
    let right = read_logs_from_json(File::open("./expected/failed_logs.json").unwrap());
    LogsAsserter::new(right, left).assert();
}

#[test]
pub fn test_assert_logs_eq() {
    let logs = read_logs_from_json(File::open("./expected/logs.json").unwrap());
    LogsAsserter::new(logs.clone(), logs).assert();
}
