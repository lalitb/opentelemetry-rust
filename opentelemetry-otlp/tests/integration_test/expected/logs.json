{
    "resourceLogs": [
      {
        "resource": {
          "attributes": [
            {
              "key": "service.name",
              "value": {
                "stringValue": "logs-integration-test"
              }
            }
          ]
        },
        "scopeLogs": [
          {
            "scope": {
              "name": "opentelemetry-log-appender",
              "version": "0.3.0"
            },
            "logRecords": [
              {
                "observedTimeUnixNano": "1715670958332465664",
                "severityNumber": 9,
                "severityText": "INFO",
                "body": {
                  "stringValue": "hello from banana. My price is 2.99."
                },
                "traceId": "",
                "spanId": "",
                "timeUnixNano": "2323232232",
                "attributes": []
              }
            ]
          }
        ]
      }
    ]
  }